Orange/classification/rules.py

```diff
@@ -0,0 +1,947 @@
+
+
+class CustomGeneralValidator(Validator):
+    """
+    Discard rules that
+
+    - cover less than the minimum required number of examples,
+    - offer no additional advantage compared to their parent rule,
+    - are too complex.
+    """
+    def __init__(self, max_rule_length=5, minimum_covered_examples=1):
+        self.max_rule_length = max_rule_length
+        self.minimum_covered_examples = minimum_covered_examples
+
+    def validate_rule(self, rule):
+        rule_target_covered = (rule.curr_class_dist[rule.target_class]
+                               if rule.target_class is not None
+                               else np.sum(rule.curr_class_dist))
+
+        return (rule_target_covered >= self.minimum_covered_examples and
+                rule_length(rule) <= self.max_rule_length and
+                (not np.array_equal(rule.curr_class_dist,
+                                    rule.parent_rule.curr_class_dist)
+                 if rule.parent_rule is not None else True))
+
+
+class LRSValidator(Validator):
+    """
+    To test significance, calculate the likelihood ratio statistic. It
+    provides an information-theoretic measure of the (non-commutative)
+    distance between two distributions. Under suitable assumptions, it
+    can be shown that the statistic is distributed approximately as
+    Chi^2 probability distribution with n-1 degrees of freedom. As the
+    score lowers, the apparent regularity is more likely observed due to
+    chance.
+    """
+    def __init__(self, alpha=1.0):
+        self.alpha = alpha
+
+    def validate_rule(self, rule):
+        if self.alpha >= 1.0 or rule.parent_rule is None:
+            return True
+
+        x = rule.curr_class_dist.astype(dtype=np.float)
+        y = rule.parent_rule.curr_class_dist.astype(dtype=np.float)
+        if rule.target_class is not None:
+            x = np.array([x[rule.target_class],
+                          np.sum(x) - x[rule.target_class]])
+            y = np.array([y[rule.target_class],
+                          np.sum(y) - y[rule.target_class]])
+
+        lrs = likelihood_ratio_statistic(x, y)
+        return (lrs > 0 and
+                chisqprob(lrs, len(rule.curr_class_dist) - 1) <= self.alpha)
+
+
+class SearchAlgorithm:
+    """
+    Implement a procedure to maneuver through the search space towards a
+    better solution, guided by the search heuristics.
+    """
+    def select_candidates(self, rules):
+        """
+        Select candidate rules to be further specialised.
+
+        Parameters
+        ----------
+        rules : Rule list
+            An ordered list of rules (best come first).
+
+        Returns
+        -------
+        candidate_rules : Rule list
+            Chosen rules.
+        rules : Rule list
+            Rules not chosen, i.e. the remainder.
+        """
+        raise NotImplementedError("descendants of SearchAlgorithm must "
+                                  "overload method 'select_candidates'")
```

Comment:
In future, you can just `raise NotImplementedError`. This is for programmers anyway; raising this exception makes it obvious the method needs to be reimplemented.